discord_bots/cogs/common.py

```diff
@@ -0,0 +1,574 @@
+                    player_display_names: list[str] = (
+                        [player.name for player in players_in_queue]
+                        if players_in_queue
+                        else []
+                    )
+                    newline = "\n"  # Escape sequence (backslash) not allowed in expression portion of f-string prior to Python 3.12
+                    embed.add_field(
+                        name=queue_title_str,
+                        value=(
+                            "> \n** **"  # weird hack to create an empty quote
+                            if not player_display_names
+                            else f">>> {newline.join(player_display_names)}"
+                        ),
+                        inline=True,
+                    )
+                    if i == rotation_queues_len - 1 and i >= 5 and i % 3 == 2:
+                        # embeds are allowed 3 "columns" per "row"
+                        # to line everything up nicely when there's >= 5 queues and only one "column" slot left, we add a blank
+                        embed.add_field(name="", value="", inline=True)
+                    if queue.id in games_by_queue:
+                        game: InProgressGame
+                        for game in games_by_queue[queue.id]:
+                            ipg_embed = await create_in_progress_game_embed(
+                                session, game, interaction.guild
+                            )
+                            ipg_embeds.append(ipg_embed)
+            await interaction.followup.send(
+                embeds=[embed] + ipg_embeds,
+            )
+            # await interaction.channel.send(
+            #     embeds=[embed] + ipg_embeds,
+            # )
+
+    @stats.autocomplete("category_name")
+    async def category_autocomplete_with_user_id(
+        self, interaction: Interaction, current: str
+    ):
+        # useful for when you want to filter the categories based on the ones the author has games played in
+        choices = []
+        session: SQLAlchemySession
+        with Session() as session:
+            result = (
+                session.query(Category.name, PlayerCategoryTrueskill.player_id)
+                .join(PlayerCategoryTrueskill)
+                .filter(PlayerCategoryTrueskill.player_id == interaction.user.id)
+                .order_by(Category.name)
+                .limit(25)  # discord only supports up to 25 choices
+                .all()
+            )
+            category_names: list[str] = [r[0] for r in result] if result else []
+            for name in category_names:
+                if current in name:
+                    choices.append(
+                        app_commands.Choice(
+                            name=name,
+                            value=name,
+                        )
+                    )
+        return choices
+
+    async def queue_autocomplete(self, interaction: Interaction, current: str):
```

Comment:
I see some of these e.g. queue_autocomplete are defined identically in multiple cogs (this one doesn't appear to have any references). Could these be combined to a central place?

Reply:
Good catch, removed the unused autocomplete function from `common.py`.

You're correct that these are replicated a couple times, and 95% of them are identical. 
My only concern with externalising them from each individual cog was the decorator requirement.
As the autocomplete function is decorated with the slash command function (and so the slash command must be in scope), and not the other way around, it seemed tricky to move them elsewhere.

Creating a new `AutocompleteCommands` cog (or similar) doesn't compile, as you cannot use the function from another cog as the decorator inside the new cog.
Similarly, creating a new file (that is not a cog), and just loading the defs into there (similar to `commands` or `utils` compiles, but does not resolve the results when running the slash command it is attempting to autocomplete.

TLDR;
The autocomplete function seems to need to be in the same scope as the function it is referring to.

Reply:
The specific def referenced is removed in d4871a3